src/test/scala/org/constellation/consensus/RoundTest.scala

```diff
@@ -1,148 +1,346 @@
+
+    round ! LightTransactionsProposal(
+      roundId,
+      FacilitatorId(daoId),
+      Seq(tx3.hash),
+      Seq(),
+      Seq()
+    )
+    round ! LightTransactionsProposal(
+      roundId,
+      facilitatorId1,
+      Seq(tx1.hash),
+      Seq(),
+      Seq()
+    )
+    round ! LightTransactionsProposal(
+      roundId,
+      facilitatorId2,
+      Seq(tx2.hash),
+      Seq(),
+      Seq()
+    )
+
+    dataResolver.resolveTransactions(*, List(), None)(3 seconds, dao) was called
+  }
+
+  test("it should resolve missing messages on union block proposals step") {
+    dao.readyPeers shouldReturn Map()
+
+    dao.messageService.contains(*) shouldReturn IO.pure(false)
+
+    dataResolver.resolveMessages(*, *, *)(3 seconds, dao) shouldReturn IO.pure(None)
+
+    round ! LightTransactionsProposal(
+      roundId,
+      FacilitatorId(daoId),
+      Seq(tx3.hash),
+      Seq(),
+      Seq()
+    )
+    round ! LightTransactionsProposal(
+      roundId,
+      facilitatorId1,
+      Seq(tx1.hash),
+      Seq(),
+      Seq()
+    )
+    round ! LightTransactionsProposal(
+      roundId,
+      facilitatorId2,
+      Seq(tx2.hash),
+      Seq("msg-hash"),
+      Seq()
+    )
+
+    dataResolver.resolveMessages("msg-hash", List(), None)(3 seconds, dao) was called
+  }
+
+  test("it should broadcast union block proposal after union block proposals step") {
+    round ! LightTransactionsProposal(
+      roundId,
+      FacilitatorId(daoId),
+      Seq(tx3.hash),
+      Seq(),
+      Seq()
+    )
+    round ! LightTransactionsProposal(
+      roundId,
+      facilitatorId1,
+      Seq(tx1.hash),
+      Seq(),
+      Seq()
+    )
+    round ! LightTransactionsProposal(
+      roundId,
+      facilitatorId2,
+      Seq(tx2.hash),
+      Seq(),
+      Seq()
+    )
+
+    round.underlyingActor.passToParentActor(any[BroadcastUnionBlockProposal]) was called
+  }
+
+  test("it should combine all received union block proposals") {
+    val cmd1 = UnionBlockProposal(
+      roundId,
+      facilitatorId1,
+      mock[CheckpointBlock]
+    )
+
+    val cmd2 = UnionBlockProposal(
+      roundId,
+      facilitatorId2,
+      mock[CheckpointBlock]
+    )
+
+    round ! cmd1
+    round ! cmd2
+
+    within(3 seconds) {
+      expectNoMessage
+      round.underlyingActor.checkpointBlockProposals.size shouldBe 2
     }
   }
 
+  test("it should send ResolveMajorityCheckpointBlock to self when received all union block proposals") {
+    round ! UnionBlockProposal(
+      roundId,
+      FacilitatorId(daoId),
+      cb1,
+    )
+    round ! UnionBlockProposal(
+      roundId,
+      facilitatorId1,
+      cb2,
+    )
+    round ! UnionBlockProposal(
+      roundId,
+      facilitatorId2,
+      cb3
+    )
+
+    round.underlyingActor.resolveMajorityCheckpointBlock() was called
+  }
+
+  test("it should resolve majority checkpoint block") {
+    round ! mock[ResolveMajorityCheckpointBlock]
+
+    // TODO: verify accepted block
```

Comment:
?

Reply:
as I wrote you on slack, most likely the implementation will be changed and right now requires lots of mocking. it's better to write the unit tests along with changed implementation instead of doing it twice. anyway, I'll take a look when I'll have a moment waiting for some questions, reviews or sth